archive/100-theorems-list/90_stirlings_formula.lean

```diff
@@ -0,0 +1,874 @@
+  rw sq,
+  linarith,
+end
+
+/-- For any `n`, we have `bn 1 - bn n ≤ 1/4` -/
+lemma bn_bounded_aux : ∀ (n : ℕ), bn 1 - bn n.succ ≤ 1 / 4 :=
+begin
+  let bn' : (ℕ → ℝ) := λ (k : ℕ), bn k.succ,
+  intro n,
+  calc
+  bn 1 - bn n.succ = bn' 0 - bn' n : rfl
+   ... = ∑ k in range n, (bn' k - bn' (k + 1))          : by rw ← (sum_range_sub' bn' n)
+   ... = ∑ k in range n, (bn k.succ - bn k.succ.succ)   : rfl
+   ... ≤ ∑ k in range n, 1 / (4 * k.succ * k.succ.succ) : sum_le_sum (λ k, λ hk, bn_sub_bn_succ k)
+   ... = ∑ k in range n, (1 / 4) * (1 / (k.succ * k.succ.succ)) :
+   begin
+     have hi : ∀ (k : ℕ), (1 : ℝ) / (4 * k.succ * k.succ.succ) =
+      (1 / 4) * (1 / (k.succ * k.succ.succ)) :=
+     begin
+       intro k,
+       norm_cast,
+       field_simp,
+       simp only [one_div, inv_inj],
+       ring_nf,
+     end,
+    refine sum_congr rfl (λ k, λ hk, hi k),
+   end
+   ... = 1 / 4 * ∑ k in range n, 1 / (k.succ * k.succ.succ) : by rw mul_sum
+   ... ≤ 1 / 4 * 1 :
+   begin
+     refine (mul_le_mul_left _).mpr _,
+     { exact div_pos one_pos four_pos },
+     { convert rat.cast_le.mpr (partial_sum_consecutive_reciprocals n),
+       rw rat_cast_sum,
+       push_cast,
+       exact rat.cast_one.symm },
+   end
+   ... = 1 / 4 : by rw mul_one,
+end
+
+/-- The sequence `bn` is bounded below by `3/4 - 1/2 * log 2`  for `n ≥ 1`. -/
+lemma bn_bounded_by_constant : ∀ (n : ℕ), 3 / (4 : ℝ) - 1 / 2 * log 2 ≤ bn n.succ :=
+begin
+  intro n,
+  calc
+  bn n.succ ≥ bn 1 - 1 / 4 : sub_le.mp (bn_bounded_aux n)
+   ... = (log ((1 : ℕ).factorial) - 1 / 2 * log (2 * (1 : ℕ)) - (1 : ℕ) *
+          log ((1 : ℕ) / (exp 1))) - 1 / 4 : by rw bn_formula 0
+   ... = 0 - 1 / 2 * log 2 - log (1 / (exp 1)) - 1 / 4 :
+      by simp only [factorial_one, cast_one, log_one, one_div, mul_one, log_inv, log_exp, mul_neg]
+   ... = -1 / 2 * log 2 - log (1 / (exp 1)) - 1 / 4 : by ring
+   ... = -1 / 2 * log 2 + 1 - 1 / 4  : by simp only [one_div, log_inv, log_exp, sub_neg_eq_add]
+   ... = -1 / 2 * log 2 + 3 / 4      : by ring
+   ... = 3 / (4 : ℝ) - 1 / 2 * log 2 : by ring,
+end
+
+/-- The sequence `bn` is bounded below. -/
+lemma bn_has_lower_bound : (lower_bounds (set.range (λ (k : ℕ), bn k.succ))).nonempty :=
+begin
+   use 3 / (4 : ℝ) - 1 / 2 * log 2,
+   intros,
+   rw lower_bounds,
+   rw [set.mem_set_of_eq],
+   simp only [set.mem_range, forall_apply_eq_imp_iff', forall_exists_index],
+   exact bn_bounded_by_constant,
+end
+
+/--
+Any sequence `bn` of real numbers that is monotone decreasing and bounded below has
+a limit in the real numbers.
+-/
+lemma monotone_convergence (bn : ℕ → ℝ) (h_sd : ∀ (n m : ℕ), n ≤ m → bn m ≤ bn n)
+  (h_bounded : (lower_bounds (set.range bn)).nonempty) : ∃ (m : ℝ), tendsto bn at_top (𝓝 m) :=
+begin
+ use Inf (set.range bn),
+ exact tendsto_at_top_is_glb h_sd (real.is_glb_Inf (set.range bn)
+   (set.range_nonempty bn) h_bounded),
+end
+
+/-- The sequence `an` is positive for `n > 0`  -/
+lemma an'_pos : ∀ (n : ℕ), 0 < an n.succ :=
```

Comment:
You can use `$` to get rid of some parentheses. Also, why not put `(n : ℕ)` before the semicolon?

Reply:
I didn't know how to make this a term proof with the `(n : ℕ)` before the semicolon.
Will check out how to use `$` here.

Reply:
Put the `(n : ℕ)` before the semicolon. 
Not familiar with using `$`.

Reply:
This is a maximally `$`-ified proof:
```
lemma stirling_seq'_pos (n : ℕ) : 0 < stirling_seq n.succ :=
div_pos (cast_pos.mpr $ factorial_pos n.succ : (0:ℝ) < _) $
  mul_pos (real.sqrt_pos.mpr $ mul_pos two_pos $ cast_pos.mpr $ succ_pos n) $
    pow_pos (div_pos (cast_pos.mpr $ succ_pos n) $ exp_pos 1) n.succ
```

Reply:
By the way, currently there are some indentations in `log_stirling_seq_bounded_aux`, `log_stirling_seq_bounded_by_constant`, `monotone_convergence` and `stirling_seq'_bdd_below` using an odd number of spaces, which should probably be fixed (always two spaces from one indentation level to the next).

Reply:
Thanks, that helped! I got rid of some triple parentheses and the spacing should be fixed now.